### DISCUSSION TITLE
Integrate VictoriaMetrics external service as metrics for Kiali -- client_error: client error: 401

### DISCUSSION THREAD

**Discussion by ivankovnatsky** [kiali/kiali#5060] [Q&A] [answered]
I do not think this is a bug, rather a configuration issue on my side, probably.

We're using terraform to deploy kiali:

```terraform
resource "helm_release" "kiali" {
  repository       = "https://kiali.org/helm-charts"
  name             = "kiali-operator"
  chart            = "kiali-operator"
  namespace        = "kiali-operator"
  create_namespace = true

  values = [
    <<-EOT
cr:
  create: true
  namespace: "istio-system"

  spec:
    auth:
      strategy: "anonymous"

    external_services:
      prometheus:
        auth:
          password: ${var.vm_token}
          username: ${var.vm_user}
        url: "https://metrics.asdf.zxcv:2222"
EOT
  ]

  depends_on = [
    null_resource.istio,
  ]
}
```

The logs show up:

```
2022-05-10T09:15:37Z INF [Prom Cache] Enabled
2022-05-10T09:15:38Z ERR Failed to fetch Prometheus configuration: client_error: client error: 401
2022-05-10T09:15:38Z ERR Failed to fetch Prometheus flags: client_error: client error: 401
```

Judging from the issue here: https://github.com/kiali/kiali/issues/3631. It seems like people have configured VictoriaMetrics as the source of metrics for Kiali no problem.

The VictoriaMetrics UI shows up this page: 
![image](https://user-images.githubusercontent.com/75213/167604943-71f9e101-3b3b-4115-8304-872afc494a68.png)

so I'm thinking that maybe I need to specify some particular `/path/`.

Update: When deploying a sample prometheus app with quick-start way:

```
kubectl apply -f https://raw.githubusercontent.com/istio/istio/release-1.13/samples/addons/prometheus.yaml
```
the graph works as expected and kiali's basic functionality.

**Comment by lucasponce** [comment]
I was confused at first as Kiali internally uses another UI component called Victory.

So, "VictoriaMetrics" is a "Prometheus" compatible API server:
![image](https://user-images.githubusercontent.com/1662329/167607289-0d42aea9-cb14-4bdd-8ee9-2edf1c06b4df.png)

I moved the issue to "enhancement" and "help wanted" as those integrations are far from the close priorities but if someone from the community would like to contribute, we'd be happy to help provide assistance and review.

**Comment by ivankovnatsky** [comment]
Thanks.

**Comment by nrfox** [comment]
> I do not think this is a bug, rather a configuration issue on my side, probably.

@ivankovnatsky moving this to a discussion as I think you're right that it's more of a configuration issue.

**Comment by nrfox** [comment, marked as answer]
@ivankovnatsky have you tried setting the `external_services.prometheus.auth.type` field to `basic`? https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.external_services.prometheus.auth.type

From the issue you linked earlier, it seems like the integration should work even if there are some warnings. 401s usually indicate an issue with auth configuration.

**Reply by ivankovnatsky** [reply]
@nrfox awesome, that worked, thanks.

**Comment by ivankovnatsky** [comment]
Can't seem to mark answered, can you guys please help.

**Reply by israel-hdez** [reply]
I marked the answer.
